### PR TITLE
[WIP] Add priority and fairness

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
Started drafting doc for kubernetes/enhancements#1040

This is #15899 but targeted for 1.17 instead of 1.16; see that older PR for comments, not all of which have been addressed yet.

### CODE REVIEW

**Comment by sftim** [comment]
Draft documentation
/retitle [WIP] Add priority and fairness

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+exceeding its length limit; rather, the request is immediately
+rejected.  Second, each API server is configured with a limit on how
+long a request can wait in a queue.  If that limit expires before the
+request can be handled then the request is rejected.  When rejecting
+requests, the API server responds using HTTP status code 429 "Too Many
+Requests".
+
+## Configuration
+
+Configuration is mainly through API objects of kinds
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]:
```suggestion
You can configure request prioritization management through two kinds of API object:
```

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+number of FIFO queues.  Whenever that priority level's number of
+requests currently being handled drops below the concurrency limit,
+the API server draws requests to begin handling from these queues in a
+fair way.
+
+## Flow schema
+
+To determine how to control a request, an API server first classifies
+that request to a _flow_ according to a configured collection of _flow
+schemas_.  The flow schemas are considered in order, according to each
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: >  The flow schemas are considered in order, according to each flow schema's _matching precedence_ (which is a number).

I recommend writing in the active voice, for example:
```markdown
Each FlowSchema is a set of rules along with a _matching precendence_. The
API server evaluates the FlowSchemas in order, sorted based on based on their
matching precedence (an integer).
A flow schema also includes the name of the priority level to use for
requests that match.
```
(is it an integer? I guessed. The reader would want to know.)

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+flow schema's _matching precedence_ (which is a number).  A flow
+schema has rules that describe which requests match that schema.
+Matching takes into account (a) the requested operation&mdash;as
+identified by the HTTP request verb and URI&mdash;and (b) if the request
+arrived through a secured port then the authenticated user attributes
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]:
```suggestion
arrived through a secured port then the authenticated user attributes,
```
I think it's important for a reader to understand what “authenticated user” means here; does it apply to clients that authenticate as a ServiceAccount? Does it just mean the humans who operate the cluster? Automated clients external to their cluster?

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+construed) to which access is managed.  In particular, each API server
+applies some prioritization and fairness in its handling of requests.
+This includes not only kube-apiservers but aggregated API servers too.
+
+Prioritization and fairness applies only to requests that are not
+_long-running_.  The nature of the request, not the size of its
+reply, determines whether a request is long-running.  The requests
+considered to be long running are
+[WATCHES](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes),
+[EXECs](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#exec),
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]:
```suggestion
[EXECs](/docs/reference/generated/kubectl/kubectl-commands#exec),
```

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+and
+[fetching of logs](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#logs)
+(with or without "following").
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: Is that list complete?

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+{{% /capture %}}
+
+
+{{% capture body %}}
+
+## Enabling API priority and fairness
+
+API priority and fairness is an alpha feature, so it is disabled by
+default.  To enable API priority and fairness you need to do two
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: [A list may help out here](https://github.com/kubernetes/website/pull/15899/files#r318084170) (@kbhawkey)

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+things: turn on the RequestManagement feature gate (e.g., by including
+the `--feature-gates RequestManagement=true` flag when starting your
+API servers) and enable the flowcontrol.apiserver.k8s.io/v1alpha1 API
+version (e.g., by including the `--runtime-config
+flowcontrol.apiserver.k8s.io/v1alpha1=true` flag when starting your
+`kube-apiserver`s).
+
+Enabling the feature gate replaces the "max in flight" filter, which
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: TODO: document the existing _max in flight_ filter.

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+general version of the "max in flight" filter.
+
+## Priority level
+
+Each API server has an aggregate concurrency limit and divides
+it&mdash;without examining the request load&mdash;among several
+priority levels.  The priority levels operate independently of each
+other.  That is, the API server classifies a given request to exactly
+one priority level and, for a given priority level, handles at most
+the corresponding number of requests at any given time.  One special
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: Maybe start “One special level” on a new paragraph, or reword to make the meaning clear.

The priority level is named `exempt`, right? That's not just a description. If someone localizes these docs it needs to be clear that `exempt` is an opaque term and not a concept being introduced.

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+priority level&mdash;used for meta-requests and other top-priority
+traffic&mdash;imposes no limit.  This priority level and the requests
+mapped to it are described as _exempt_, because of the lack of
+concurrency limitation.
+
+For a non-exempt request, if it is not allowed to be handled
+immediately then the API server will either reject the request or hold
+it in a queue for a while.  Each non-exempt priority level has a
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: [for a while—can you be more specific?](https://github.com/kubernetes/website/pull/15899/files#r318114677) (@kbhawkey)

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md (hunk 3 above)
Comment (sftim) [reply content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: Also:
> what I was trying to say is wrong; this needs to be factually corrected (not just reworded).

(@MikeSpreitzer)

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+## Queuing
+
+An API server first classifies a request to a flow, and thus a
+corresponding priority level, and if the priority level is not exempt
+then the API server assigns the request to a queue using a technique
+called "shuffle sharding".  This technique starts with hashing the
+flow identifier, and then uses the hash value as a source of entropy
+to pseudo-randomly pick a small subset of the queues, and finally
+assigns the request to one of the shortest queues in that subset.  The
+priority level declares the size of the subset.
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: [what does “declares” mean here?](https://github.com/kubernetes/website/pull/15899/files#r318787061) (@yliaog)

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+API server immediately begins handling the request if that would not
+violate the priority level's concurrency limit; otherwise the queue
+holds the request until it can be executed or is evicted.
+
+The API server applies two criteria to limit queue occupancy.  First,
+each non-exempt priority level has a queue length limit.  The
+API server will not add a request to a queue that is already at or
+exceeding its length limit; rather, the request is immediately
+rejected.  Second, each API server is configured with a limit on how
+long a request can wait in a queue.  If that limit expires before the
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: [worth to mention a request is rejected only when the priority level's concurrency limit is exceeded](https://github.com/kubernetes/website/pull/15899/files#r318790307) (@yliaog)

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+requests, the API server responds using HTTP status code 429 "Too Many
+Requests".
+
+## Configuration
+
+Configuration is mainly through API objects of kinds
+`PriorityLevelConfiguration` and `FlowSchema` in the
+`flowcontrol.apiserver.k8s.io` API group.
+
+The API server's aggregate concurrency limit is the sum of the two
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: @kbhawkey [requested](https://github.com/kubernetes/website/pull/15899/files#r318120741) an example

File: content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+each non-exempt priority level has a queue length limit.  The
+API server will not add a request to a queue that is already at or
+exceeding its length limit; rather, the request is immediately
+rejected.  Second, each API server is configured with a limit on how
+long a request can wait in a queue.  If that limit expires before the
+request can be handled then the request is rejected.  When rejecting
+requests, the API server responds using HTTP status code 429 "Too Many
+Requests".
+
+## Configuration
```
Comment (sftim) [inline content/en/docs/concepts/cluster-administration/api-priority-and-fairness.md]: @jaypipes [requested](https://github.com/kubernetes/website/pull/15899/files#r315866472) YAML-format examples

File: content/en/docs/reference/access-authn-authz/controlling-access.md [hunk 14]
```diff
@@ -61,9 +61,22 @@ While Kubernetes uses `usernames` for access control decisions and in request lo
 it does not have a `user` object nor does it store usernames or other information about
 users in its object store.
 
+## Prioritization and fairness
+
+After the request is authenticated as coming from a specific user, the
+request is subject to prioritization and fairness if the
+RequestManagement feature is enabled.  _Some more words need to go
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/controlling-access.md]: :warning: Just highlighting to reviewers that this is a TODO

File: content/en/docs/reference/access-authn-authz/using-api-priority-and-fairness.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+API servers can apply prioritization and fairness in how incoming
+requests get access to the server's limited capacity to handle requests.
+{{% /capture %}}
+
+{{% capture body %}}
+This page should reprise the Motivation and Proposal sections of
+[the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md)
+in user-facing terms.
+
+Beware: the KEP is still being actively revised
```
Comment (sftim) [inline content/en/docs/reference/access-authn-authz/using-api-priority-and-fairness.md]: :warning: Just highlighting to reviewers that this is incomplete

**Comment by daminisatya** [comment]
@MikeSpreitzer 

This is a kind reminder since we're approaching the deadline for PR to be "Ready for Review" by Nov 15, 2019. Kindly update this PR.

**Comment by MikeSpreitzer** [comment]
Sorry, this did not get done in time.  We will try for release 1.18

**Comment by sftim** [comment]
This didn't make v1.17
@MikeSpreitzer I'm going to close this. Please do feel free to open a new PR based on the same commits / revised commits, targeting a future release.
/close

**Pull request closed without merging**